clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -741,61 +767,86 @@ boolean reconcile() {
                 revokedPartitions
         );
 
-        CompletableFuture<Void> revocationResult;
-        if (!revokedPartitions.isEmpty()) {
-            revocationResult = revokePartitions(revokedPartitions);
+        // Commit offsets if auto-commit enabled before reconciling a new assignment. Request will
+        // be retried until it succeeds, fails with non-retriable error, or timer expires.
+        CompletableFuture<Void> commitResult;
+
+        if (commitRequestManager.autoCommitEnabled()) {
+            // TODO: review auto commit time boundary. This will be effectively bounded by the
+            //  rebalance timeout.
+            commitResult = commitRequestManager.autoCommitAllConsumedNow(Optional.of(time.timer(Long.MAX_VALUE)));
         } else {
-            revocationResult = CompletableFuture.completedFuture(null);
-            // Reschedule the auto commit starting from now (new assignment received without any
-            // revocation).
-            commitRequestManager.resetAutoCommitTimer();
+            commitResult = CompletableFuture.completedFuture(null);
         }
 
-        // Future that will complete when the full reconciliation process completes (revocation
-        // and assignment, executed sequentially).
-        CompletableFuture<Void> reconciliationResult =
+        // Execute commit -> onPartitionsRevoked -> onPartitionsAssigned.
+        commitResult.whenComplete((commitReqResult, commitReqError) -> {
```

Comment:
Nesting futures inside futures is something we want to avoid. Can you try to structure it something like this?


```
commitResult
      .handle((commitReqResult, commitReqError) -> {
          // log errors
      })
      .thenCompose(__ -> {
          // revoke partitions
      })
      .thenCompose(__ -> {
          // reconcile
      })
      .whenComplete((result, error) -> {
          markReconciliationCompleted();
          ...
```

This would also be a good opportunity to extract things into separate methods

Reply:
Agree. I simplified it a bit by extracting things into separate methods. The thenCompose approach doest not fit well in some cases here where the sequence should continue on failure (ex. commit->revocation). I do have a follow-up already where all this will be greatly simplified btw, with [KAFKA-15835](https://issues.apache.org/jira/browse/KAFKA-15835), where the intention is to group the commit->revocation->assign possibly in a single event.